tests/beta/interfaces/test_tape_autograd.py

```diff
@@ -732,3 +730,559 @@ def cost(x, device):
+)
+class TestQNode:
+    """Same tests as above, but this time via the QNode interface!"""
+
+    def test_execution_no_interface(self, dev_name, diff_method):
+        """Test execution works without an interface"""
+        if diff_method == "backprop":
+            pytest.skip("Test does not support backprop")
+
+        dev = qml.device(dev_name, wires=1)
+
+        @qnode(dev, interface=None)
+        def circuit(a):
+            qml.RY(a, wires=0)
+            qml.RX(0.2, wires=0)
+            return expval(qml.PauliZ(0))
+
+        a = np.array(0.1, requires_grad=True)
+
+        res = circuit(a)
+
+        assert circuit.qtape.interface == None
+
+        # without the interface, the tape simply returns an array of results
+        assert isinstance(res, np.ndarray)
+        assert res.shape == (1,)
+        assert isinstance(res[0], float)
+
+        # without the interface, the tape is unable to deduce
+        # trainable parameters
+        assert circuit.qtape.trainable_params == {0, 1}
+
+        # gradients should cause an error
+        with pytest.raises(TypeError, match="must be real number, not ArrayBox"):
+            qml.grad(circuit)(a)
+
+    def test_execution_with_interface(self, dev_name, diff_method):
+        """Test execution works with the interface"""
+        if diff_method == "backprop":
+            pytest.skip("Test does not support backprop")
+
+        dev = qml.device(dev_name, wires=1)
+
+        @qnode(dev, interface="autograd")
+        def circuit(a):
+            qml.RY(a, wires=0)
+            qml.RX(np.array(0.2, requires_grad=False), wires=0)
+            return expval(qml.PauliZ(0))
+
+        a = np.array(0.1, requires_grad=True)
+        circuit(a)
+
+        assert circuit.qtape.interface == "autograd"
+
+        # the tape is able to deduce trainable parameters
+        assert circuit.qtape.trainable_params == {0}
+
+        # gradients should work
+        grad = qml.grad(circuit)(a)
+        assert len(grad) == 1
+        assert isinstance(grad[0], np.ndarray)
+        assert grad[0].shape == tuple()
+
+    def test_interface_swap(self, dev_name, diff_method, tol):
+        """Test that the autograd interface can be applied to a QNode
+        with a pre-existing interface"""
+        tf = pytest.importorskip("tensorflow", minversion="2.1")
+
+        if diff_method == "backprop":
+            pytest.skip("Test does not support backprop")
+
+        dev = qml.device(dev_name, wires=1)
+
+        @qnode(dev, interface="tf")
+        def circuit(a):
+            qml.RY(a, wires=0)
+            qml.RX(np.array(0.2, requires_grad=False), wires=0)
+            return expval(qml.PauliZ(0))
+
+        a = tf.Variable(0.1, dtype=tf.float64)
+
+        with tf.GradientTape() as tape:
+            res_tf = circuit(a)
+
+        grad_tf = tape.gradient(res_tf, a)
+
+        # switch to autograd interface
+        circuit.to_autograd()

```

Comment:
Just curious: would there be use cases where changing interfaces could come in handy?

Reply:
Yep, there might be some places where we'll want to programmatically apply interfaces to a provided QNode, that might already have an interface.